homeassistant/components/device_tracker/owntracks.py

```diff
@@ -60,47 +67,101 @@ def owntracks_event_update(topic, payload, qos):
             data = json.loads(payload)
         except ValueError:
             # If invalid JSON
-            logging.getLogger(__name__).error(
+            _LOGGER.error(
                 'Unable to parse payload as JSON: %s', payload)
             return
 
         if not isinstance(data, dict) or data.get('_type') != 'transition':
             return
 
-        # check if in "home" fence or other zone
-        location = ''
+        if data['desc'].lower() == 'home':
+            location = STATE_HOME
+        else:
+            location = data['desc']
+
+        dev_id, kwargs = _parse_see_args(topic, data)
+
         if data['event'] == 'enter':
+            zone = hass.states.get("zone.{}".format(location))
+            with LOCK:
+                if zone is not None:
+                    kwargs['location_name'] = location
 
-            if data['desc'].lower() == 'home':
-                location = STATE_HOME
-            else:
-                location = data['desc']
+                    regions = REGIONS_ENTERED[dev_id]
+                    if location not in regions:
+                        regions.append(location)
+                    _LOGGER.info("Enter region %s", location)
+                    _get_gps_from_zone(kwargs, zone)
+
+                see(**kwargs)
 
         elif data['event'] == 'leave':
-            location = STATE_NOT_HOME
+            current_location = hass.states.get(
+                "device_tracker.{}".format(dev_id)).state
+
+            regions = REGIONS_ENTERED[dev_id]
+            if location in regions:
+                regions.remove(location)
+            new_region = regions[-1] if regions else None
+
+            if not new_region:
+                _LOGGER.info("Exit from %s to GPS", location)
+                see(**kwargs)
+
+                # if gps location didn't set new state, force to away
+                current_location = hass.states.get(
```

Comment:
When would it happen that it didn't set a new state and how would we be able to detect this without having to call `see` twice, because that seems like a hack.